msprime/intervals.py

```diff
@@ -33,83 +33,232 @@ class RateMap:
     """
     A class mapping a numeric value to a set of adjacent intervals along the
     genome.
+
+    :param list position: A list of :math:`n+1` positions, starting at 0, and ending
+        in the sequence length over which the RateMap will apply.
+    :param list rate: A list of :math:`n` positive rates that apply between each
+        position.
+    :param float start_position: When returning a mean_rate over the genome, ignore the
+        part of the map from zero up to ``start_position``, as long as the rate is 0 in
+        that region. If ``None`` (default), the ``start_position`` is 0.
+    :param float end_position: When returning a mean_rate over the genome, ignore the
+        part of the map from ``end_position`` onwards, as long as the rate is 0 in that
+        region. If ``None`` (default), the ``end_position`` is the end of the map.
     """
 
-    def __init__(self, position, rate, map_start=0):
-        # We take copies to make sure there are no unintended consequences
-        # later when a user modifies the arrays in this map.
-        self.position = np.array(position, dtype=float, copy=True)
-        self.rate = np.array(rate, dtype=float, copy=True)
-        self.map_start = map_start
-        size = len(self.rate)
-        if size < 1:
+    def __init__(
+        self,
+        position,
+        rate,
+        *,
+        start_position=None,
+        end_position=None,
+    ):
+        # Taking copies ensures there are no unintended consequences later if a user
+        # modifies the input arrays.
+
+        # Save positions into a read-only array
+        self._position = np.array(position, dtype=float, copy=True)
+        size = len(self.position)
```

Comment:
I find it confusing (and potentially error-prone; you'd normally assume the constructor as competed successfully when writing a property accessor) to use both the private variable and the public property within the constructor. I think it'd be most appropriate to just use ``self._position``, etc within the constructor

Reply:
Yes, I wasn't sure which to use. I'll switch to `self._XXX`